node/node.go

```diff
@@ -0,0 +1,452 @@
+	dir         string
+	l           net.Listener
+	wk          consensus.PrivateKey
+
+	mux TreeMux
+	srv http.Server
+
+	cleanupFuncs []func() error
+}
+
+// NewNode creates a new, empty node without any components. The components are
+// then added through the various exposed methods. For a more detailed
+// description take a look at the Node's type documentation.
+func NewNode(apiAddr, apiPassword, dir string, uiHandler http.Handler, wk consensus.PrivateKey) *Node {
+	// Start listening.
+	l, err := net.Listen("tcp", apiAddr)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer l.Close()
+
+	mux := TreeMux{
+		H:   uiHandler,
+		Sub: make(map[string]http.Handler),
+	}
+
+	// Overwrite APIAddr now that we know the exact addr from the listener.
+	return &Node{
+		apiAddr:     "http://" + l.Addr().String(),
+		apiPassword: apiPassword,
+		auth:        jape.BasicAuth(apiPassword),
+		dir:         dir,
+		l:           l,
+		mux:         mux,
+		srv:         http.Server{Handler: mux},
+		wk:          wk,
+	}
+}
+
+// APIAddress returns the node's API address.
+func (c *Node) APIAddress() string {
+	return c.apiAddr
+}
+
+// Close shuts down the API as well as the components that were created.
+// Remote components that were added instead of created will remain active.
+func (c *Node) Close() error {
+	if err := c.srv.Shutdown(context.Background()); err != nil {
+		return err
+	}
+	var errs error
+	for _, f := range c.cleanupFuncs {
+		err := f()
+		if err == nil {
+			continue
+		} else if errs == nil {
+			errs = err
+		} else {
+			errs = fmt.Errorf("%w; %s", errs, err.Error())
+		}
+	}
+	return errs
+}
+
+// Serve starts serving the API. This is blocking and can be interrupted by
+// calling 'Close'.
+func (c *Node) Serve() error {
+	err := c.srv.Serve(c.l)
+	if errors.Is(err, net.ErrClosed) {
+		return nil // ignore
+	}
+	return err
+}
+
+// AddBus adds an already running, remote bus to the node.
+func (c *Node) AddBus(busAddr, busPassword string) error {
+	if c.busAddr != nil || c.busPassword != nil {
+		return errors.New("cluster already contains a bus")
+	}
+	// Set bus address and password to remote bus.
+	*c.busAddr = busAddr
+	*c.busPassword = busPassword
+	return nil
+}
+
+// CreateBus creates a new bus.
+func (c *Node) CreateBus(bootstrap bool, gatewayAddr string) error {
+	if c.busAddr != nil || c.busPassword != nil {
+		return errors.New("cluster already contains a bus")
+	}
+	b, cleanup, err := newBus(BusConfig{
+		Bootstrap:   bootstrap,
+		GatewayAddr: gatewayAddr,
+	}, c.dir, c.wk)
+	if err != nil {
+		return err
+	}
+	// Set bus address and password to local bus.
+	c.busAddr = &c.apiAddr
+	c.busPassword = &c.apiPassword
+	c.cleanupFuncs = append(c.cleanupFuncs, cleanup)
+	c.bus = b
+	c.mux.Sub["/api/store"] = &TreeMux{H: c.auth(bus.NewServer(b))}
+	return nil
+}
+
+// AddWorker adds an already running, remote worker to the node.
+func (c *Node) AddWorker(workerAddr, workerPassword string) error {
```

Comment:
what's the reason for `Node` to expose both `CreateWorker` and `AddWorker`

Reply:
`CreateWorker` "creates" a worker in the node which means the node itself serves the API routes for the worker.
The node is also responsible for the cleanup of that worker.

`AddWorker` adds a remote worker which is not managed by the node. 

Same for `AddBus` and `CreateBus`. The only reason the autopilot doesn't have that is because a node doesn't really need to connect to an existing autopilot.